File: app/models/analysis.py
"""
分析相关数据模型
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum
from bson import ObjectId
from .user import PyObjectId


class AnalysisStatus(str, Enum):
    """分析状态枚举"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class BatchStatus(str, Enum):
    """批次状态枚举"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    PARTIAL_SUCCESS = "partial_success"
    FAILED = "failed"
    CANCELLED = "cancelled"


class AnalysisParameters(BaseModel):
    """分析参数模型

    研究深度说明：
    - 快速: 1级 - 快速分析 (2-4分钟)
    - 基础: 2级 - 基础分析 (4-6分钟)
    - 标准: 3级 - 标准分析 (6-10分钟，推荐)
    - 深度: 4级 - 深度分析 (10-15分钟)
    - 全面: 5级 - 全面分析 (15-25分钟)
    """
    market_type: str = "A股"
    analysis_date: Optional[datetime] = None
    research_depth: str = "标准"  # 默认使用3级标准分析（推荐）
    selected_analysts: List[str] = Field(default_factory=lambda: ["market", "fundamentals", "news", "social"])
    custom_prompt: Optional[str] = None
    include_sentiment: bool = True
    include_risk: bool = True
    language: str = "zh-CN"
    # 模型配置
    quick_analysis_model: Optional[str] = "qwen-turbo"
    deep_analysis_model: Optional[str] = "qwen-max"


class AnalysisResult(BaseModel):
    """分析结果模型"""
    analysis_id: Optional[str] = None
    summary: Optional[str] = None
    recommendation: Optional[str] = None
    confidence_score: Optional[float] = None
    risk_level: Optional[str] = None
    key_points: List[str] = Field(default_factory=list)
    detailed_analysis: Optional[Dict[str, Any]] = None
    charts: List[str] = Field(default_factory=list)
    tokens_used: int = 0
    execution_time: float = 0.0
    error_message: Optional[str] = None


class AnalysisTask(BaseModel):
    """分析任务模型"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    task_id: str = Field(..., description="任务唯一标识")
    batch_id: Optional[str] = None
    user_id: PyObjectId
    symbol: str = Field(..., description="6位股票代码")
    stock_code: Optional[str] = Field(None, description="股票代码(已废弃,使用symbol)")
    stock_name: Optional[str] = None
    status: AnalysisStatus = AnalysisStatus.PENDING

    progress: int = Field(default=0, ge=0, le=100, description="任务进度 0-100")
    
    # 时间戳
    created_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    
    # 执行信息
    worker_id: Optional[str] = None
    parameters: AnalysisParameters = Field(default_factory=AnalysisParameters)
    result: Optional[AnalysisResult] = None
    
    # 重试机制
    retry_count: int = 0
    max_retries: int = 3
    last_error: Optional[str] = None
    
    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}


class AnalysisBatch(BaseModel):
    """分析批次模型"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    batch_id: str = Field(..., description="批次唯一标识")
    user_id: PyObjectId
    title: str = Field(..., description="批次标题")
    description: Optional[str] = None
    status: BatchStatus = BatchStatus.PENDING
    
    # 任务统计
    total_tasks: int = 0
    completed_tasks: int = 0
    failed_tasks: int = 0
    cancelled_tasks: int = 0
    progress: int = Field(default=0, ge=0, le=100, description="整体进度 0-100")
    
    # 时间戳
    created_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    
    # 配置参数
    parameters: AnalysisParameters = Field(default_factory=AnalysisParameters)
    
    # 结果摘要
    results_summary: Optional[Dict[str, Any]] = None
    
    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}


class StockInfo(BaseModel):
    """股票信息模型"""
    symbol: str = Field(..., description="6位股票代码")
    code: Optional[str] = Field(None, description="股票代码(已废弃,使用symbol)")
    name: str = Field(..., description="股票名称")
    market: str = Field(..., description="市场类型")
    industry: Optional[str] = None
    sector: Optional[str] = None
    market_cap: Optional[float] = None
    price: Optional[float] = None
    change_percent: Optional[float] = None


# API请求/响应模型

class SingleAnalysisRequest(BaseModel):
    """单股分析请求"""
    symbol: Optional[str] = Field(None, description="6位股票代码")
    stock_code: Optional[str] = Field(None, description="股票代码(已废弃,使用symbol)")
    parameters: Optional[AnalysisParameters] = None

    def get_symbol(self) -> str:
        """获取股票代码(兼容旧字段)"""
        return self.symbol or self.stock_code or ""


class BatchAnalysisRequest(BaseModel):
    """批量分析请求"""
    title: str = Field(..., description="批次标题")
    description: Optional[str] = None
    symbols: Optional[List[str]] = Field(None, min_items=1, max_items=10, description="股票代码列表（最多10个）")
    stock_codes: Optional[List[str]] = Field(None, min_items=1, max_items=10, description="股票代码列表(已废弃,使用symbols，最多10个)")
    parameters: Optional[AnalysisParameters] = None

    def get_symbols(self) -> List[str]:
        """获取股票代码列表(兼容旧字段)"""
        return self.symbols or self.stock_codes or []


class AnalysisTaskResponse(BaseModel):
    """分析任务响应"""
    task_id: str
    batch_id: Optional[str]
    symbol: str
    stock_code: Optional[str] = None  # 兼容字段
    stock_name: Optional[str]
    status: AnalysisStatus
    progress: int
    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    result: Optional[AnalysisResult]


class AnalysisBatchResponse(BaseModel):
    """分析批次响应"""
    batch_id: str
    title: str
    description: Optional[str]
    status: BatchStatus
    total_tasks: int
    completed_tasks: int
    failed_tasks: int
    progress: int
    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    parameters: AnalysisParameters


class AnalysisHistoryQuery(BaseModel):
    """分析历史查询参数"""
    status: Optional[AnalysisStatus] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    symbol: Optional[str] = None
    stock_code: Optional[str] = None  # 兼容字段
    batch_id: Optional[str] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)

    def get_symbol(self) -> Optional[str]:
        """获取股票代码(兼容旧字段)"""
        return self.symbol or self.stock_code

File: frontend/src/types/analysis.ts
// 分析状态枚举
export enum AnalysisStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

// 批次状态枚举
export enum BatchStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  PARTIAL_SUCCESS = 'partial_success',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

// 分析参数
export interface AnalysisParameters {
  market_type: 'A股' | '美股' | '港股'
  analysis_date?: string
  research_depth: '快速' | '基础' | '标准' | '深度' | '全面'
  selected_analysts: string[]
  custom_prompt?: string
  include_charts: boolean
  language: 'zh-CN' | 'en-US'
}

// 分析结果
export interface AnalysisResult {
  analysis_id: string
  summary?: string
  recommendation?: string
  confidence_score?: number
  risk_level?: string
  key_points: string[]
  detailed_analysis?: Record<string, any>
  charts: string[]
  tokens_used: number
  execution_time: number
  error_message?: string
}

// 分析任务
export interface AnalysisTask {
  id: string
  task_id: string
  batch_id?: string
  user_id: string
  symbol?: string  // 主字段：品种代码
  stock_code?: string  // 兼容字段（已废弃）
  stock_name?: string
  status: AnalysisStatus
  priority: number
  progress: number
  
  // 时间戳
  created_at: string
  started_at?: string
  completed_at?: string
  
  // 执行信息
  worker_id?: string
  parameters: AnalysisParameters
  result?: AnalysisResult
  
  // 重试机制
  retry_count: number
  max_retries: number
  last_error?: string
}

// 分析批次
export interface AnalysisBatch {
  id: string
  batch_id: string
  user_id: string
  title: string
  description?: string
  status: BatchStatus
  
  // 任务统计
  total_tasks: number
  completed_tasks: number
  failed_tasks: number
  cancelled_tasks: number
  progress: number
  
  // 时间戳
  created_at: string
  started_at?: string
  completed_at?: string
  
  // 配置参数
  parameters: AnalysisParameters
  
  // 结果摘要
  results_summary?: Record<string, any>
}

// 品种信息（统一字段名）
export interface InstrumentInfo {
  // 基础信息
  symbol: string           // 品种代码（如 CU2608.SHF）
  code?: string            // 简码（如 CU）
  full_symbol?: string     // 完整代码
  name: string             // 品种名称
  market: string           // 市场类型
  industry?: string        // 行业/品类
  area?: string
  board?: string
  exchange?: string        // 交易所

  // 市值信息（亿元）
  total_mv?: number        // 总市值
  circ_mv?: number         // 流通市值

  // 财务指标
  pe?: number
  pb?: number
  pe_ttm?: number
  pb_mrq?: number
  roe?: number

  // 交易数据
  close?: number
  pct_chg?: number
  amount?: number
  turnover_rate?: number
  volume_ratio?: number

  // 技术指标
  ma20?: number
  rsi14?: number
  kdj_k?: number
  kdj_d?: number
  kdj_j?: number
  dif?: number
  dea?: number
  macd_hist?: number
}

// 向后兼容别名（已废弃，请使用 InstrumentInfo）
export type StockInfo = InstrumentInfo

// 单股分析请求
export interface SingleAnalysisRequest {
  symbol?: string  // 主字段：品种代码
  stock_code?: string  // 兼容字段（已废弃）
  parameters?: AnalysisParameters
}

// 批量分析请求
export interface BatchAnalysisRequest {
  title: string
  description?: string
  symbols?: string[]  // 主字段：品种代码列表
  stock_codes?: string[]  // 兼容字段（已废弃）
  parameters?: AnalysisParameters
}

// 分析任务响应
export interface AnalysisTaskResponse {
  task_id: string
  batch_id?: string
  symbol?: string  // 主字段：品种代码
  stock_code?: string  // 兼容字段（已废弃）
  stock_name?: string
  status: AnalysisStatus
  progress: number
  created_at: string
  started_at?: string
  completed_at?: string
  result?: AnalysisResult
}

// 分析批次响应
export interface AnalysisBatchResponse {
  batch_id: string
  title: string
  description?: string
  status: BatchStatus
  total_tasks: number
  completed_tasks: number
  failed_tasks: number
  progress: number
  created_at: string
  started_at?: string
  completed_at?: string
  parameters: AnalysisParameters
}

// 分析历史查询参数
export interface AnalysisHistoryQuery {
  status?: AnalysisStatus
  start_date?: string
  end_date?: string
  symbol?: string  // 主字段：品种代码
  stock_code?: string  // 兼容字段（已废弃）
  batch_id?: string
  page: number
  page_size: number
}

// 分析历史响应
export interface AnalysisHistoryResponse {
  tasks: AnalysisTask[]
  batches: AnalysisBatch[]
  total: number
  page: number
  page_size: number
  has_more: boolean
}

// 任务进度信息
export interface TaskProgress {
  task_id: string
  status: AnalysisStatus
  progress: number
  message?: string
  updated_at: string
}

// 批次进度信息
export interface BatchProgress {
  batch_id: string
  status: BatchStatus
  total_tasks: number
  completed_tasks: number
  failed_tasks: number
  progress: number
  updated_at: string
}

// 分析统计信息
export interface AnalysisStats {
  total_analyses: number
  successful_analyses: number
  failed_analyses: number
  success_rate: number
  average_execution_time: number
  total_tokens_used: number
  daily_analyses: number
  monthly_analyses: number
}

// 队列状态信息
export interface QueueStatus {
  pending: number
  processing: number
  completed: number
  failed: number
  queue_size: number
}

// 用户队列状态
export interface UserQueueStatus {
  pending: number
  processing: number
  concurrent_limit: number
  available_slots: number
}

// 分析报告
export interface AnalysisReport {
  id: string
  task_id: string
  batch_id?: string
  title: string
  content: string
  format: 'html' | 'pdf' | 'markdown'
  created_at: string
  download_url?: string
}
